File: python/jeefs/image.py
# SPDX-License-Identifier: (GPL-2.0+ or Apache-2.0)
"""Whole-image build and parse for the JEEFS filesystem (filesystem-v1.md).

The library performs no I/O: :func:`build_image` returns the bytes the
environment writes to the medium, :func:`parse_image` consumes the bytes
the environment read. Semantics follow docs/format/filesystem-v1.md
(28-byte file header, absolute nextFileAddress, double CRC, the
fs_version gate at header offset 10) and header-v4.md (fs_version is
stamped when a filesystem is written, refreshing the header CRC).
"""

from __future__ import annotations

import binascii
import struct
from dataclasses import dataclass, field

from .constants_generated import (
    EEPROM_DEVICE_ID_FILENAME,
    EEPROM_FILE_NAME_LENGTH,
    EEPROM_FS_VERSION,
    EEPROM_FS_VERSION_OFFSET,
)
from .header import EEPROMHeaderV3, EEPROMHeaderV4, detect_version

DEVICE_ID_FILENAME = EEPROM_DEVICE_ID_FILENAME

_FHDR = 28  # sizeof(JEEFSFileHeaderv1)
_HEADER_SIZES = {1: 512, 2: 256, 3: 256, 4: 256}
_MAX_DATA = 32767  # INT16_MAX, filesystem-v1.md constraints
_MAX_IMAGE = 65535  # uint16_t addressing


@dataclass(frozen=True)
class ImageFile:
    """One file of the chain: a name (<= 15 chars) and its payload."""

    name: str
    data: bytes


@dataclass
class ParsedImage:
    """The parse_image result.

    ``header`` is a typed model for v3/v4 headers and ``None`` for the
    obsolete v1/v2 layouts (``header_bytes`` always carries the raw
    header). ``files`` is empty when ``fs_version`` is 0 — the file area
    of such an image is empty regardless of content.
    """

    version: int
    header_bytes: bytes
    header: EEPROMHeaderV3 | None
    fs_version: int
    files: list[ImageFile] = field(default_factory=list)


def _seal_file_header(name: str, data: bytes, next_addr: int) -> bytes:
    raw = bytearray(_FHDR)
    encoded = name.encode("ascii")
    raw[0 : len(encoded)] = encoded
    struct.pack_into("<H", raw, 16, len(data))
    struct.pack_into("<I", raw, 18, binascii.crc32(data) & 0xFFFFFFFF)
    struct.pack_into("<H", raw, 22, next_addr)
    struct.pack_into("<I", raw, 24, binascii.crc32(bytes(raw[:24])) & 0xFFFFFFFF)
    return bytes(raw)


def build_image(
    header: EEPROMHeaderV3, files: list[ImageFile | tuple[str, bytes]], image_size: int
) -> bytes:
    """Build a complete EEPROM image: board header + JEEFS file chain.

    The header is serialized via its own ``to_bytes`` and stamped with
    ``fs_version = 1`` (the image carries a — possibly empty —
    filesystem), refreshing the header CRC; the caller's object is not
    mutated. Files are laid out contiguously in the given order with one
    exception: the reserved ``device.id`` file always takes the first
    slot (filesystem-v1.md). The free space is zero-filled.

    Raises ValueError on invalid names/data, duplicates, an image size
    outside the uint16 range or too small for header + chain (capacity).
    """
    if not 0 < image_size <= _MAX_IMAGE:
        raise ValueError(f"image_size {image_size} outside 1..{_MAX_IMAGE}")

    normalized: list[ImageFile] = []
    for f in files:
        item = f if isinstance(f, ImageFile) else ImageFile(f[0], f[1])
        if not 0 < len(item.name) <= EEPROM_FILE_NAME_LENGTH:
            raise ValueError(f"file name {item.name!r} must be 1..{EEPROM_FILE_NAME_LENGTH} chars")
        if not 0 < len(item.data) <= _MAX_DATA:
            raise ValueError(f"file {item.name!r} data must be 1..{_MAX_DATA} bytes")
        normalized.append(item)
    names = [f.name for f in normalized]
    if len(set(names)) != len(names):
        raise ValueError("duplicate file names in the chain")

    # The reserved device-identity file always occupies the first slot.
    normalized.sort(key=lambda f: f.name != DEVICE_ID_FILENAME)

    header_bytes = bytearray(header.to_bytes())
    header_bytes[EEPROM_FS_VERSION_OFFSET] = EEPROM_FS_VERSION
    struct.pack_into("<I", header_bytes, 252, binascii.crc32(bytes(header_bytes[:252])) & 0xFFFFFFFF)

    chain = bytearray()
    offset = len(header_bytes)
    for i, f in enumerate(normalized):
        end = offset + _FHDR + len(f.data)
        next_addr = 0 if i == len(normalized) - 1 else end
        chain += _seal_file_header(f.name, f.data, next_addr)
        chain += f.data
        offset = end

    total = len(header_bytes) + len(chain)
    if total > image_size:
        raise ValueError(f"capacity exceeded: header + chain = {total} bytes, image_size = {image_size}")

    return bytes(header_bytes) + bytes(chain) + b"\x00" * (image_size - total)


def parse_image(data: bytes) -> ParsedImage:
    """Parse a complete EEPROM image back into header and files.

    Applies the reader rules of filesystem-v1.md: the fs_version byte
    gates the file area (0 = no filesystem, files ignored; unknown =
    error), every visited file header must pass its headerCrc32 before
    any field is trusted, links must be exactly contiguous, and every
    file's data CRC is verified.

    Raises ValueError when no valid board header is present, on an
    unknown fs_version, or on a corrupted chain.
    """
    version = detect_version(data)
    if version is None:
        raise ValueError("no valid board header (magic/version mismatch)")
    header_size = _HEADER_SIZES[version]
    if len(data) < header_size:
        raise ValueError(f"image shorter than the v{version} header")

    header_bytes = bytes(data[:header_size])
    header: EEPROMHeaderV3 | None = None
    if version == 3:
        header = EEPROMHeaderV3.from_bytes(header_bytes)
    elif version == 4:
        header = EEPROMHeaderV4.from_bytes(header_bytes)

    fs_version = data[EEPROM_FS_VERSION_OFFSET]
    result = ParsedImage(version, header_bytes, header, fs_version)
    if fs_version == 0:
        return result  # no filesystem: the file area is empty by contract
    if fs_version != EEPROM_FS_VERSION:
        raise ValueError(f"unsupported fs_version {fs_version}")

    offset = header_size
    while offset != 0 and offset + _FHDR <= len(data):
        raw = data[offset : offset + _FHDR]
        if raw[0] in (0x00, 0xFF):
            break  # unwritten slot: end of chain
        stored_hcrc = struct.unpack("<I", raw[24:28])[0]
        if binascii.crc32(raw[:24]) & 0xFFFFFFFF != stored_hcrc:
            raise ValueError(f"file header CRC mismatch at offset {offset}")
        if raw[EEPROM_FILE_NAME_LENGTH] != 0:
            raise ValueError(f"unterminated file name at offset {offset}")
        name = raw[:16].split(b"\x00")[0].decode("ascii")
        data_size, stored_crc, next_addr = struct.unpack("<HIH", raw[16:24])
        if data_size == 0 or data_size == 0xFFFF:
            raise ValueError(f"corrupted dataSize at offset {offset}")
        end = offset + _FHDR + data_size
        if end > len(data):
            raise ValueError(f"file {name!r} runs past the image")
        if next_addr == 0xFFFF:
            next_addr = 0  # erased link terminates like 0
        if next_addr != 0 and (next_addr != end or end + _FHDR > len(data)):
            raise ValueError(f"broken chain link at offset {offset}")
        payload = bytes(data[offset + _FHDR : end])
        if binascii.crc32(payload) & 0xFFFFFFFF != stored_crc:
            raise ValueError(f"file {name!r} data CRC mismatch")
        result.files.append(ImageFile(name, payload))
        offset = next_addr

    return result

File: rust/jeefs-header/src/image.rs
// SPDX-License-Identifier: (GPL-2.0+ or Apache-2.0)
//! Whole-image build and parse for the JEEFS filesystem (filesystem-v1.md).
//!
//! The library performs no I/O: [`build_image`] returns the bytes the
//! environment writes to the medium, [`parse_image`] consumes the bytes the
//! environment read. The rules mirror the C reader in `src/jeefs.c` and the
//! Python port; cross-language conformance is locked by the byte-identity
//! of the committed goldens (ctest `image_build_rs_matches_golden*`).

use crate::generated::{DEVICE_ID_FILENAME, FILE_NAME_LENGTH, FS_VERSION, FS_VERSION_OFFSET};
use crate::header::{detect_version, header_size};
use std::string::String;
use std::vec::Vec;

const FHDR: usize = 28; // sizeof(JEEFSFileHeaderv1)
const MAX_DATA: usize = 32767; // INT16_MAX: the C API's int16_t byte counts
const MAX_IMAGE: usize = 65535; // uint16_t addressing

/// One file of the chain: a name (<= 15 byte-range chars) and its payload.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ImageFile {
    pub name: String,
    pub data: Vec<u8>,
}

/// The [`parse_image`] result. `files` is empty when `fs_version` is 0 —
/// the file area of such an image is empty regardless of content.
/// `unreadable` mirrors the C API's split between listing and reading:
/// names whose chain entry is valid but whose payload `EEPROM_ReadFile`
/// would refuse (data-CRC mismatch or dataSize above INT16_MAX); such
/// files still appear in `files` with the payload as stored.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedImage {
    pub version: u8,
    pub header_bytes: Vec<u8>,
    pub fs_version: u8,
    pub files: Vec<ImageFile>,
    pub unreadable: Vec<String>,
}

/// Errors of the image surface. Chain corruption carries the offset of
/// the offending file header.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ImageError {
    /// No valid board header (magic/version mismatch or short buffer).
    NoHeader,
    /// The board header fails its own CRC — a failed CRC is a failed
    /// header: no field is interpreted (checked on raw bytes before the
    /// typed model, ahead of every other rule).
    HeaderCrc,
    /// The fs_version byte names a layout this build does not know.
    FsVersion(u8),
    /// A file header fails validation at the given offset.
    Chain(usize),
    /// A file name is empty, too long, or carries NUL / non-byte chars.
    BadName,
    /// A payload is empty or above INT16_MAX bytes.
    BadData,
    /// Two files share a name.
    DuplicateName,
    /// image_size is outside 1..=65535 or cannot hold header + chain.
    Capacity,
}

impl core::fmt::Display for ImageError {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            ImageError::NoHeader => write!(f, "no valid board header"),
            ImageError::HeaderCrc => write!(f, "board header CRC mismatch"),
            ImageError::FsVersion(v) => write!(f, "unsupported fs_version {v}"),
            ImageError::Chain(off) => write!(f, "corrupted file chain at offset {off}"),
            ImageError::BadName => write!(f, "invalid file name"),
            ImageError::BadData => write!(f, "invalid file payload size"),
            ImageError::DuplicateName => write!(f, "duplicate file names"),
            ImageError::Capacity => write!(f, "image size cannot hold header and chain"),
        }
    }
}

impl std::error::Error for ImageError {}

fn crc32(data: &[u8]) -> u32 {
    crc32fast::hash(data)
}

fn seal_file_header(name: &str, data: &[u8], next_addr: u16) -> [u8; FHDR] {
    let mut raw = [0u8; FHDR];
    for (i, ch) in name.chars().enumerate() {
        raw[i] = ch as u8; // validated byte-range chars only
    }
    raw[16..18].copy_from_slice(&(data.len() as u16).to_le_bytes());
    raw[18..22].copy_from_slice(&crc32(data).to_le_bytes());
    raw[22..24].copy_from_slice(&next_addr.to_le_bytes());
    let hc = crc32(&raw[..24]);
    raw[24..28].copy_from_slice(&hc.to_le_bytes());
    raw
}

/// Build a complete EEPROM image: board header + JEEFS file chain.
///
/// `header` is a prepared board header (its CRC is refreshed here after
/// `fs_version = 1` is stamped into the copy — the image carries a,
/// possibly empty, filesystem; the caller's buffer is untouched). Files
/// are laid out contiguously in the given order with one exception: the
/// reserved `device.id` file always takes the first slot
/// (filesystem-v1.md). The free space is zero-filled.
pub fn build_image(
    header: &[u8],
    files: &[(&str, &[u8])],
    image_size: usize,
) -> Result<Vec<u8>, ImageError> {
    if image_size == 0 || image_size > MAX_IMAGE {
        return Err(ImageError::Capacity);
    }
    let version = detect_version(header).ok_or(ImageError::NoHeader)?;
    let hdr_size = header_size(version).ok_or(ImageError::NoHeader)?;
    if header.len() < hdr_size {
        return Err(ImageError::NoHeader);
    }

    for (name, data) in files {
        let n = name.chars().count();
        if n == 0 || n > FILE_NAME_LENGTH || name.chars().any(|c| c == '\0' || c as u32 > 0xFF) {
            return Err(ImageError::BadName);
        }
        if data.is_empty() || data.len() > MAX_DATA {
            return Err(ImageError::BadData);
        }
    }
    let mut seen = std::collections::HashSet::new();
    for (name, _) in files {
        if !seen.insert(*name) {
            return Err(ImageError::DuplicateName);
        }
    }

    // The reserved device-identity file always occupies the first slot;
    // the sort is stable, so the rest keep their order.
    let mut ordered: Vec<&(&str, &[u8])> = files.iter().collect();
    ordered.sort_by_key(|(name, _)| *name != DEVICE_ID_FILENAME);

    let mut image = Vec::with_capacity(image_size);
    image.extend_from_slice(&header[..hdr_size]);
    image[FS_VERSION_OFFSET] = FS_VERSION as u8;
    let coverage = hdr_size - 4;
    let hc = crc32(&image[..coverage]).to_le_bytes();
    image[coverage..hdr_size].copy_from_slice(&hc);

    let mut offset = hdr_size;
    for (i, (name, data)) in ordered.iter().enumerate() {
        let end = offset + FHDR + data.len();
        if end > image_size {
            return Err(ImageError::Capacity);
        }
        let next = if i == ordered.len() - 1 {
            0
        } else {
            end as u16
        };
        image.extend_from_slice(&seal_file_header(name, data, next));
        image.extend_from_slice(data);
        offset = end;
    }
    if image.len() > image_size {
        return Err(ImageError::Capacity);
    }
    image.resize(image_size, 0);
    Ok(image)
}

/// Parse a complete EEPROM image back into header bytes and files.
///
/// The reader rules of filesystem-v1.md, exactly as the C iterator
/// applies them — with one stricter gate: this API returns header field
/// bytes, so the board-header CRC is verified on the raw bytes before
/// anything else is interpreted (a failed CRC is a failed header).
pub fn parse_image(data: &[u8]) -> Result<ParsedImage, ImageError> {
    let version = detect_version(data).ok_or(ImageError::NoHeader)?;
    let hdr_size = header_size(version).ok_or(ImageError::NoHeader)?;
    if data.len() < hdr_size {
        return Err(ImageError::NoHeader);
    }

    let coverage = hdr_size - 4;
    let stored = u32::from_le_bytes(data[coverage..hdr_size].try_into().unwrap());
    if crc32(&data[..coverage]) != stored {
        return Err(ImageError::HeaderCrc);
    }

    let fs_version = data[FS_VERSION_OFFSET];
    let mut parsed = ParsedImage {
        version,
        header_bytes: data[..hdr_size].to_vec(),
        fs_version,
        files: Vec::new(),
        unreadable: Vec::new(),
    };
    if fs_version == 0 {
        return Ok(parsed); // no filesystem: the file area is empty by contract
    }
    if fs_version != FS_VERSION as u8 {
        return Err(ImageError::FsVersion(fs_version));
    }

    let mut offset = hdr_size;
    while offset != 0 && offset + FHDR <= data.len() {
        let raw = &data[offset..offset + FHDR];
        if raw[0] == 0x00 || raw[0] == 0xFF {
            break; // unwritten slot: end of chain
        }
        let stored_hcrc = u32::from_le_bytes(raw[24..28].try_into().unwrap());
        if crc32(&raw[..24]) != stored_hcrc {
            return Err(ImageError::Chain(offset));
        }
        if raw[FILE_NAME_LENGTH] != 0 {
            return Err(ImageError::Chain(offset));
        }
        let data_size = u16::from_le_bytes(raw[16..18].try_into().unwrap()) as usize;
        let stored_crc = u32::from_le_bytes(raw[18..22].try_into().unwrap());
        let mut next = u16::from_le_bytes(raw[22..24].try_into().unwrap()) as usize;
        if data_size == 0 || data_size == 0xFFFF {
            return Err(ImageError::Chain(offset));
        }
        let end = offset + FHDR + data_size;
        if end > data.len() {
            return Err(ImageError::Chain(offset));
        }
        if next == 0xFFFF {
            next = 0; // erased link terminates like 0 (RFC #14)
        }
        if next != 0 && (next != end || end + FHDR > data.len()) {
            return Err(ImageError::Chain(offset));
        }

        let name: String = raw[..16]
            .iter()
            .take_while(|&&b| b != 0)
            .map(|&b| b as char)
            .collect();
        let payload = data[offset + FHDR..end].to_vec();
        if data_size > MAX_DATA || crc32(&payload) != stored_crc {
            parsed.unreadable.push(name.clone());
        }
        parsed.files.push(ImageFile {
            name,
            data: payload,
        });
        offset = next;
    }

    Ok(parsed)
}
